### PR TITLE
test: collect fuller logs for CI tests

### PR DESCRIPTION
**Pull request by BlaineEXE**
A few of the nightly tests are failing often, and we were missing logs for many things that would be helpful. Rewrite the log collector to collect all pod logs from cluster and system namespaces, to describe all common k8s resources, and to describe all CRs in those namespaces.

Signed-off-by: Blaine Gardner <blaine.gardner@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Comment by BlaineEXE** [comment]
On a local test install, the files output look like below:
```
test
├── ceph-osd-dump.txt
├── ceph-report.txt
├── ceph-status.txt
├── dmesg.txt
├── journalctl.txt
└── namespace-rook-ceph
    ├── all-wide.txt
    ├── all-yaml.txt
    ├── cephblockpools-describe--builtin-nfs.txt
    ├── cephblockpools-describe--device-health-metrics.txt
    ├── cephblockpools-describe--replicapool.txt
    ├── cephclusters-describe--my-cluster.txt
    ├── cephfilesystems-describe--myfs.txt
    ├── cephnfses-describe--my-nfs.txt
    ├── cephobjectstores-describe--my-store.txt
    ├── cephobjectstoreusers-describe--my-user.txt
    ├── daemonset-describe--csi-cephfsplugin.txt
    ├── daemonset-describe--csi-rbdplugin.txt
    ├── daemonset-list.txt
    ├── deployment-describe--csi-cephfsplugin-provisioner.txt
    ├── deployment-describe--csi-rbdplugin-provisioner.txt
    ├── deployment-describe--rook-ceph-mds-myfs-a.txt
    ├── deployment-describe--rook-ceph-mds-myfs-b.txt
    ├── deployment-describe--rook-ceph-mgr-a.txt
    ├── deployment-describe--rook-ceph-mon-a.txt
    ├── deployment-describe--rook-ceph-nfs-my-nfs-a.txt
    ├── deployment-describe--rook-ceph-operator.txt
    ├── deployment-describe--rook-ceph-osd-0.txt
    ├── deployment-describe--rook-ceph-osd-1.txt
    ├── deployment-describe--rook-ceph-osd-2.txt
    ├── deployment-describe--rook-ceph-rgw-my-store-a.txt
    ├── deployment-describe--rook-ceph-tools.txt
    ├── deployment-list.txt
    ├── job-describe--rook-ceph-osd-prepare-minikube.txt
    ├── job-list.txt
    ├── logs--csi-cephfsplugin-gkvsz.txt
    ├── logs--csi-cephfsplugin-provisioner-6f54f6c477-wff7x.txt
    ├── logs--csi-rbdplugin-c85cf.txt
    ├── logs--csi-rbdplugin-provisioner-6d765b47d5-4p75l.txt
    ├── logs--rook-ceph-mds-myfs-a-786ffd5c69-c4mcg.txt
    ├── logs--rook-ceph-mds-myfs-b-7f6dc5c669-4gsd8.txt
    ├── logs--rook-ceph-mgr-a-f656f9cfc-x54cr.txt
    ├── logs--rook-ceph-mon-a-54b5f88bdf-j4hpd.txt
    ├── logs--rook-ceph-nfs-my-nfs-a-5f4779d5fd-kcvgr.txt
    ├── logs--rook-ceph-operator-77f744ffb7-5ccq8.txt
    ├── logs--rook-ceph-operator-b77b97c9c-f78hq.txt
    ├── logs--rook-ceph-osd-0-95687fd94-997g8.txt
    ├── logs--rook-ceph-osd-1-57dfc6f8f4-kkghj.txt
    ├── logs--rook-ceph-osd-2-649d454ff7-tfr9j.txt
    ├── logs--rook-ceph-osd-prepare-minikube--1-k2sjp.txt
    ├── logs--rook-ceph-rgw-my-store-a-76fbf94cff-dzgc9.txt
    ├── logs--rook-ceph-tools-555c879675-xqcb6.txt
    ├── objectbuckets-describe--obc-default-ceph-delete-bucket.txt
    ├── pod-describe--csi-cephfsplugin-gkvsz.txt
    ├── pod-describe--csi-cephfsplugin-provisioner-6f54f6c477-wff7x.txt
    ├── pod-describe--csi-rbdplugin-c85cf.txt
    ├── pod-describe--csi-rbdplugin-provisioner-6d765b47d5-4p75l.txt
    ├── pod-describe--rook-ceph-mds-myfs-a-786ffd5c69-c4mcg.txt
    ├── pod-describe--rook-ceph-mds-myfs-b-7f6dc5c669-4gsd8.txt
    ├── pod-describe--rook-ceph-mgr-a-f656f9cfc-x54cr.txt
    ├── pod-describe--rook-ceph-mon-a-54b5f88bdf-j4hpd.txt
    ├── pod-describe--rook-ceph-nfs-my-nfs-a-5f4779d5fd-kcvgr.txt
    ├── pod-describe--rook-ceph-operator-77f744ffb7-5ccq8.txt
    ├── pod-describe--rook-ceph-operator-b77b97c9c-f78hq.txt
    ├── pod-describe--rook-ceph-osd-0-95687fd94-997g8.txt
    ├── pod-describe--rook-ceph-osd-1-57dfc6f8f4-kkghj.txt
    ├── pod-describe--rook-ceph-osd-2-649d454ff7-tfr9j.txt
    ├── pod-describe--rook-ceph-osd-prepare-minikube--1-k2sjp.txt
    ├── pod-describe--rook-ceph-rgw-my-store-a-76fbf94cff-dzgc9.txt
    ├── pod-describe--rook-ceph-tools-555c879675-xqcb6.txt
    └── pod-list.txt

1 directory, 70 files
```

**Comment by BlaineEXE** [comment]
Compare this to what currently gets logged, with the note that `deploy-describe-*` files are blank because of a bug in the collector script.
```
.
├── ceph-osd-dump.txt
├── ceph-report.txt
├── ceph-status.txt
├── cephcluster.txt
├── cluster-jobs-list.txt
├── cluster-pods-list.txt
├── cluster-wide.txt
├── cluster-yaml.txt
├── deploy-describe-rook-ceph-crashcollector-fv-az270-670-log.txt
├── deploy-describe-rook-ceph-crashcollector-fv-az270-670.txt
├── deploy-describe-rook-ceph-mds-ceph-filesystem-test-a-log.txt
├── deploy-describe-rook-ceph-mds-ceph-filesystem-test-a.txt
├── deploy-describe-rook-ceph-mds-ceph-filesystem-test-b-log.txt
├── deploy-describe-rook-ceph-mds-ceph-filesystem-test-b.txt
├── deploy-describe-rook-ceph-mgr-a-log.txt
├── deploy-describe-rook-ceph-mgr-a.txt
├── deploy-describe-rook-ceph-mon-a-log.txt
├── deploy-describe-rook-ceph-mon-a.txt
├── deploy-describe-rook-ceph-osd-0-log.txt
├── deploy-describe-rook-ceph-osd-0.txt
├── deploy-describe-rook-ceph-rgw-ceph-objectstore-test-a-log.txt
├── deploy-describe-rook-ceph-rgw-ceph-objectstore-test-a.txt
├── deploy-describe-rook-ceph-tools-log.txt
├── deploy-describe-rook-ceph-tools.txt
├── dmesg.txt
├── journalctl.txt
├── lsblk.txt
├── operator-logs.txt
├── operator-pods-list.txt
├── operator-wide.txt
├── operator-yaml.txt
├── osd-prepare-describe.txt
├── osd-prepare-logs.txt
├── osd-removal-describe.txt
├── osd-removal-logs.txt
├── pod-describe-file-test.txt
├── pod-describe-rook-ceph-crashcollector-fv-az270-670-668db79879-2zlhd.txt
├── pod-describe-rook-ceph-mds-ceph-filesystem-test-a-6dd78879fb-qv749.txt
├── pod-describe-rook-ceph-mds-ceph-filesystem-test-b-bdc66c7d8-kq5jj.txt
├── pod-describe-rook-ceph-mgr-a-54456d9b74-bsh5w.txt
├── pod-describe-rook-ceph-mon-a-846f89997c-mn8zr.txt
├── pod-describe-rook-ceph-osd-0-84769f558d-hrwjj.txt
├── pod-describe-rook-ceph-osd-prepare-fv-az270-670-qw86n.txt
├── pod-describe-rook-ceph-rgw-ceph-objectstore-test-a-84fc4d4bbd-6pq4w.txt
├── pod-describe-rook-ceph-tools-7f6598cb58-8w8bq.txt
├── rook-ceph-osd-0-logs.txt
└── rook-ceph-osd-1-logs.txt
```

File: tests/scripts/collect-logs.sh [hunk 1]
```diff
@@ -16,30 +16,43 @@ $CEPH_CMD -s > "${LOG_DIR}"/ceph-status.txt
...
+    for resource in $(kubectl -n "$NAMESPACE" get "$CRD" -o jsonpath='{.items[*].metadata.name}'); do
+      crd_main_type="${CRD%%.*}" # e.g., for cephclusters.ceph.rook.io, only use 'cephclusters'
+      kubectl -n "$NAMESPACE" get -o yaml "$CRD" "$resource" > "${NS_DIR}"/"$crd_main_type"-describe--"$resource".txt
+    done
+  done
+
+  # do simple 'get all' calls for resources we don't often want to look at
+  kubectl get all -n "$NAMESPACE" -o wide > "${NS_DIR}"/all-wide.txt
+  kubectl get all -n "$NAMESPACE" -o yaml > "${NS_DIR}"/all-yaml.txt
+done # NAMESPACE
```
Comment (subhamkrai) [inline tests/scripts/collect-logs.sh:54]:
```suggestion
done
```
?

**Review by subhamkrai — CHANGES_REQUESTED** [review]
looks good, also checked the artifacts locally.

small lint

File: .github/workflows/integration-test-upgrade-suite.yaml [hunk 2]
```diff
@@ -94,12 +94,11 @@ jobs:
...
     - name: Artifact
       uses: actions/upload-artifact@v2
-      if: failure()
       with:
         name: ceph-upgrade-suite-artifact-${{ matrix.kubernetes-versions }}
         path: /home/runner/work/rook/rook/tests/integration/_output/tests/
 
     - name: setup tmate session for debugging when event is PR
-      if: failure() && github.event_name == 'pull_request'
+      if: github.event_name == 'pull_request'
```
Comment (travisn) [inline .github/workflows/integration-test-upgrade-suite.yaml:102]: Was this only for temporary testing and you'll revert this commit?

File: tests/scripts/collect-logs.sh [hunk 3]
```diff
@@ -16,30 +16,43 @@ $CEPH_CMD -s > "${LOG_DIR}"/ceph-status.txt
...
+
+  # describe every one of the k8s resources in the namespace which rook commonly uses
+  for KIND in 'pod' 'deployment' 'job' 'daemonset'; do
+    kubectl -n "$NAMESPACE" get "$KIND" -o wide > "${NS_DIR}"/"$KIND"-list.txt
+    for resource in $(kubectl -n "$NAMESPACE" get "$KIND" -o jsonpath='{.items[*].metadata.name}'); do
+      kubectl -n "$NAMESPACE" describe "$KIND" "$resource" > "${NS_DIR}"/"$KIND"-describe--"$resource".txt
+
+      # collect logs for pods along the way
+      if [[ "$KIND" == 'pod' ]]; then
+        kubectl -n "$NAMESPACE" logs --all-containers "$resource" > "${NS_DIR}"/logs--"$resource".txt
```
Comment (travisn) [inline tests/scripts/collect-logs.sh:37]: Each pod will have its own log file, right?

Comment (BlaineEXE) [reply tests/scripts/collect-logs.sh:37]: Right!

File: .github/workflows/integration-test-upgrade-suite.yaml (hunk 2 above)
Comment (BlaineEXE) [reply .github/workflows/integration-test-upgrade-suite.yaml]: Yes. I'll delete the commit when the output looks good.

File: tests/scripts/collect-logs.sh (hunk 3 above)
Comment (BlaineEXE) [reply tests/scripts/collect-logs.sh:37]: The they go into `logs--*.txt` files like here: https://github.com/rook/rook/pull/9781#issuecomment-1047304151

**Pull request merged**